datafusion/spark/src/function/hash/utils.rs

```diff
@@ -0,0 +1,614 @@
+        Ok(())
+    }
+
+    fn hash_list<T>(iter: ArrayIter<T>, seed: &mut [H]) -> Result<()>
+    where
+        T: ArrayAccessor<Item = ArrayRef>,
+    {
+        try_hash_impl(iter, seed, |seed, v| {
+            let len = v.len();
+            let mut result = [seed; 1];
+            for i in 0..len {
+                let slice = v.slice(i, 1);
+                Self::hash(&slice, &mut result)?;
+            }
+            Ok(result[0])
+        })?;
+        Ok(())
+    }
+
+    fn hash_map<T>(iter: ArrayIter<T>, seed: &mut [H]) -> Result<()>
+    where
+        T: ArrayAccessor<Item = StructArray>,
+    {
+        try_hash_impl(iter, seed, |seed, v| {
+            let len = v.len();
+            let mut result = [seed; 1];
+            for i in 0..len {
+                let slice = v.slice(i, 1);
+                Self::hash(slice.column(0), &mut result)?;
+                Self::hash(slice.column(1), &mut result)?;
+            }
+            Ok(result[0])
+        })?;
+        Ok(())
+    }
+
+    fn hash_struct(arr: &StructArray, seed: &mut [H]) -> Result<()> {
+        Self::hash_arrays(arr.columns(), seed)
+    }
+
+    fn hash_arrays(arr: &[ArrayRef], seed: &mut [H]) -> Result<()> {
+        for arr in arr {
+            Self::hash(arr, seed)?;
+        }
+        Ok(())
+    }
+
+    fn hash_dictionary<T>(arr: &DictionaryArray<T>, seed: &mut [H]) -> Result<()>
+    where
+        T: ArrowDictionaryKeyType,
+    {
+        let len = arr.len();
+        if len != seed.len() {
+            return exec_err!("Array length mismatch: {} != {}", len, seed.len());
+        }
+
+        for (i, seed) in seed.iter_mut().enumerate().take(arr.len()) {
+            if arr.is_null(i) {
+                continue;
+            }
+            let mut result = [*seed; 1];
+
+            let value = arr.values().slice(i, 1);
+            let key: Arc<dyn Array> = Arc::new(arr.keys().slice(i, 1));
+
+            for j in 0..key.len() {
+                let value = value.slice(j, 1);
+                let key = key.slice(j, 1);
+                Self::hash(&key, &mut result)?;
+                Self::hash(&value, &mut result)?;
+            }
+            *seed = result[0];
+        }
+        Ok(())
+    }
+
+    fn hash_union(arr: &UnionArray, seed: &mut [H]) -> Result<()> {
+        let len = arr.len();
+        if len != seed.len() {
+            return exec_err!("Array length mismatch: {} != {}", len, seed.len());
+        }
+
+        for (i, seed) in seed.iter_mut().enumerate().take(len) {
+            let value = arr.value(i);
+            let mut result = [*seed; 1];
+            for i in 0..value.len() {
+                let elem = value.slice(i, 1);
+                Self::hash(&elem, &mut result)?;
+            }
+            *seed = result[0];
+        }
+        Ok(())
+    }
+
+    /// Computes the xxHash64 hash of the given data
+    fn hash(arr: &ArrayRef, seed: &mut [H]) -> Result<()> {
+        match arr.data_type() {
```

Comment:
I can't help feel there's a more ergonomic way to do this, perhaps using `downcast_primitive_array!()` macro? I'm still having trouble understanding all the parts of this code so I can't think of anything concrete yet.